File: exm/mway.py
class TreeNode:
    def __init__(self, t):
        self.keys = []  # Contains the keys
        self.children = []  # Contains the child nodes
        self.t = t  # Minimum degree (defines the range for number of keys)

    def is_leaf(self):
        return len(self.children) == 0

# Function to insert a key in the tree
def insert(root, key, t):
    # If root is null, tree is empty
    if root is None:
        root = TreeNode(t)
        root.keys.append(key)
        return root

    # If root is full, tree grows in height
    if len(root.keys) == 2 * t - 1:
        new_root = TreeNode(t)
        new_root.children.append(root)
        split(new_root, 0, t)
        insert_non_full(new_root, key, t)
        return new_root
    else:
        insert_non_full(root, key, t)
        return root

# Function to insert key in a non-full node
def insert_non_full(node, key, t):
    i = len(node.keys) - 1
    if node.is_leaf():
        node.keys.append(None)
        while i >= 0 and key < node.keys[i]:
            node.keys[i + 1] = node.keys[i]
            i -= 1
        node.keys[i + 1] = key
    else:
        while i >= 0 and key < node.keys[i]:
            i -= 1
        i += 1
        if len(node.children[i].keys) == 2 * t - 1:
            split(node, i, t)
            if key > node.keys[i]:
                i += 1
        insert_non_full(node.children[i], key, t)

# Function to split the child of a node
def split(parent, i, t):
    new_node = TreeNode(t)
    child = parent.children[i]
    new_node.keys = child.keys[t:]
    child.keys = child.keys[:t-1]

    if not child.is_leaf():
        new_node.children = child.children[t:]
        child.children = child.children[:t]

    parent.children.insert(i + 1, new_node)
    parent.keys.insert(i, child.keys[t-1])

# Function for in-order traversal of the tree
def in_order_traversal(node):
    if node is not None:
        for i in range(len(node.keys)):
            if len(node.children) > i:
                in_order_traversal(node.children[i])
            print(node.keys[i], end=' ')
        if len(node.children) > len(node.keys):
            in_order_traversal(node.children[-1])

# Example Usage
t = 2  # Degree of the tree
root = None
keys = [10, 20, 5, 6, 12, 30, 7, 17]
for key in keys:
    root = insert(root, key, t)

print("In-order Traversal of the M-way Search Tree:")
in_order_traversal(root)

File: exm/cpp/mway.cpp
#include <bits/stdc++.h>
using namespace std;

class TreeNode {
public:
    vector<int> keys;
    vector<TreeNode*> children;
    int t;  // Minimum degree (defines the range for number of keys)

    TreeNode(int _t, bool _isLeaf) {
        t = _t;
        keys.resize(2 * t - 1);
        children.resize(2 * t);
    }

    bool isLeaf() {
        return children[0] == nullptr;
    }
};

class MWaySearchTree {
public:
    TreeNode* root;
    int t;  // Minimum degree

    MWaySearchTree(int _t) {
        root = nullptr;
        t = _t;
    }

    void insert(int key) {
        // If root is null, create a new root
        if (root == nullptr) {
            root = new TreeNode(t, true);
            root->keys[0] = key;
            root->children[0] = nullptr;
            root->children[1] = nullptr;
        } else {
            // If root is full, then tree grows in height
            if (root->keys.size() == 2 * t - 1) {
                TreeNode* newRoot = new TreeNode(t, false);
                newRoot->children[0] = root;
                splitChild(newRoot, 0, root);
                int i = 0;
                if (newRoot->keys[0] < key) {
                    i++;
                }
                insertNonFull(newRoot->children[i], key);
                root = newRoot;
            } else {
                insertNonFull(root, key);
            }
        }
    }

private:
    void insertNonFull(TreeNode* node, int key) {
        int i = node->keys.size() - 1;

        // If this is a leaf node
        if (node->isLeaf()) {
            node->keys.resize(node->keys.size() + 1);
            while (i >= 0 && node->keys[i] > key) {
                node->keys[i + 1] = node->keys[i];
                i--;
            }
            node->keys[i + 1] = key;
        } else {
            // Find the child which is going to have the new key
            while (i >= 0 && node->keys[i] > key) {
                i--;
            }
            // Check if the found child is full
            if (node->children[i + 1]->keys.size() == 2 * t - 1) {
                splitChild(node, i + 1, node->children[i + 1]);
                if (node->keys[i + 1] < key) {
                    i++;
                }
            }
            insertNonFull(node->children[i + 1], key);
        }
    }

    void splitChild(TreeNode* parent, int i, TreeNode* child) {
        TreeNode* newChild = new TreeNode(t, child->isLeaf());
        newChild->keys.resize(t - 1);

        for (int j = 0; j < t - 1; j++) {
            newChild->keys[j] = child->keys[j + t];
        }

        if (!child->isLeaf()) {
            newChild->children.resize(t);
            for (int j = 0; j < t; j++) {
                newChild->children[j] = child->children[j + t];
            }
        }

        child->keys.resize(t - 1);
        parent->children.resize(parent->children.size() + 1);

        for (int j = parent->children.size() - 1; j >= i + 1; j--) {
            parent->children[j] = parent->children[j - 1];
        }

        parent->children[i + 1] = newChild;

        for (int j = parent->keys.size(); j >= i; j--) {
            parent->keys[j] = parent->keys[j - 1];
        }

        parent->keys[i] = child->keys[t - 1];
        parent->keys.resize(parent->keys.size() + 1);
    }
};

int main() {
    MWaySearchTree tree(3);  // Creating an M-Way Search Tree with t = 3

    // Inserting elements
    tree.insert(10);
    tree.insert(20);
    tree.insert(5);
    tree.insert(6);
    tree.insert(12);
    tree.insert(30);
    tree.insert(7);
    tree.insert(17);

    // Add code for traversal or other operations as needed

    return 0;
}
